_posts/2017-08-16-json-server.md

```diff
@@ -0,0 +1,655 @@
+
+```bash
+$ json-server generate.js
+
+  \{^_^}/ hi!
+
+  Loading generate.js
+  Done
+
+  Resources
+  http://localhost:3000/messages
+
+  Home
+  http://localhost:3000
+
+  Type s + enter at any time to create a snapshot of the database
+```
+
+And the results would be something like:
+
+```bash
+$ curl http://localhost:3000/messages
+[
+  {
+    "id": 1,
+    "name": "driver",
+    "status": "cross-platform",
+    "description": "If we connect the system, we can get to the ADP panel through the redundant PCI protocol!"
+  },
+  {
+    "id": 2,
+    "name": "monitor",
+    "status": "1080p",
+    "description": "Try to synthesize the CSS driver, maybe it will navigate the bluetooth matrix!"
+  },
+  {
+    "id": 3,
+    "name": "hard drive",
+    "status": "virtual",
+    "description": "Use the redundant SMS program, then you can compress the bluetooth port!"
+  }
+]
+```
+
+And we can still perform requests on it as we've seen in the [routes](#routes) section.
+
+<a name="custom-routes"></a>
+
+### Custom routes
+
+Imagine we are supposed to perform requests on several different endpoints on our future API, and these endpoints don't have the same URIs:
+
+```url
+/api/dashboard
+/api/groups/ducks/stats
+/auth/users
+/rpierlot/articles
+```
+
+`json-server` allows us to specify route rewrites. We can address this problem using a map that resolves the actual routes in our json schema:
+
+```json
+{
+  "/api/:view": "/:view",
+  "/api/groups/:planet/stats": "/stats?planet=:planet",
+  "/:user/articles": "/articles?author=:user",
+  "/auth/users": "/users"
+}
+```
+
+So, when we start `json-server` it shows us the route rewrites we are using :
+
+```bash
+$ json-server --watch db2.json --routes routes.json
+
+  \{^_^}/ hi!
+
+  Loading db2.json
+  Loading routes.json
+  Done
+
+  Resources
+  http://localhost:3000/users
+  http://localhost:3000/dashboard
+  http://localhost:3000/stats
+  http://localhost:3000/articles
+
+  Other routes
+  /api/:view -> /:view
+  /api/groups/:planet/stats -> /stats?planet=:planet
+  /:user/articles -> /articles?author=:user
+  /auth/users -> /users
+
+  Home
+  http://localhost:3000
+
+  Type s + enter at any time to create a snapshot of the database
+  Watching...
+```
+
+Now we can perform our custom requests to see the results:
+
+```bash
+$ curl http://localhost:3000/api/dashboard
+{
+  "visits": 3881,
+  "views": 625128,
+  "shares": 7862
+}
+
+$ curl http://localhost:3000/api/groups/ducks/stats
+[
+  {
+    "planet": "ducks",
+    "stats": {
+      "points": 5625,
+      "ships": 8
+    }
+  }
+]
+```
+
+<a name="middlewares"></a>
+
+### Middlewares
+
+In case we want to augment our `json-server` instance with a specific behavior, we have the possibility to do so using custom middlewares, these middlewares are passed to the express server the same way we would do it when developing a classic express app. In this section, we're going to explore a useful example of a feature that is usually necessary.
```

Comment:
In case we want to augment our `json-server` instance with a specific behavior, we have the possibility to do so by using custom middlewares : these middlewares are passed to the express server the same way we would do it when developing a classic express app. In this section, we're going to explore a useful example of a feature that is usually necessary.